src/EFCore.PG/Metadata/PostgresComposite.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Metadata.Internal;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Utilities;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Metadata
+{
+    public class PostgresComposite
+    {
+        readonly IAnnotatable _annotatable;
+        readonly string _annotationName;
+
+        internal PostgresComposite(IAnnotatable annotatable, string annotationName)
+        {
+            _annotatable = annotatable;
+            _annotationName = annotationName;
+        }
+
+        public static PostgresComposite GetOrAddPostgresComposite(
+            [NotNull] IMutableAnnotatable annotatable,
+            [CanBeNull] string schema,
+            [NotNull] string name,
+            [NotNull] params (string Name, string StoreType)[] fields)
+        {
+            if (FindPostgresComposite(annotatable, schema, name) is PostgresComposite CompositeType)
```

Comment:
Should we add some `Check` preconditions? Or is this always called in a known context?

Reply:
Not sure, but added anyway.